rosidl_generator_cpp/include/rosidl_generator_cpp/bounded_vector.hpp

```diff
@@ -0,0 +1,712 @@
+   *  %BoundedVector and if it is frequently used the user should
+   *  consider using std::list.
+   */
+  typename _Base::iterator
+  insert(
+    typename _Base::const_iterator __position,
+    std::initializer_list<typename _Base::value_type> __l)
+  {
+    if (size() + __l.size() > _UpperBound) {
+      throw std::length_error("Exceeded upper bound");
+    }
+    return _Base::insert(__position, __l);
+  }
+
+  /**
+   *  @brief  Inserts a number of copies of given data into the %BoundedVector.
+   *  @param  __position  A const_iterator into the %BoundedVector.
+   *  @param  __n  Number of elements to be inserted.
+   *  @param  __x  Data to be inserted.
+   *  @return  An iterator that points to the inserted data.
+   *
+   *  This function will insert a specified number of copies of
+   *  the given data before the location specified by @a position.
+   *
+   *  Note that this kind of operation could be expensive for a
+   *  %BoundedVector and if it is frequently used the user should
+   *  consider using std::list.
+   */
+  typename _Base::iterator
+  insert(
+    typename _Base::const_iterator __position,
+    typename _Base::size_type __n,
+    const typename _Base::value_type & __x)
+  {
+    if (size() + __n > _UpperBound) {
+      throw std::length_error("Exceeded upper bound");
+    }
+    return _Base::insert(__position, __n, __x);
+  }
+
+  /**
+   *  @brief  Inserts a range into the %BoundedVector.
+   *  @param  __position  A const_iterator into the %BoundedVector.
+   *  @param  __first  An input iterator.
+   *  @param  __last   An input iterator.
+   *  @return  An iterator that points to the inserted data.
+   *
+   *  This function will insert copies of the data in the range
+   *  [__first,__last) into the %BoundedVector before the location specified
+   *  by @a pos.
+   *
+   *  Note that this kind of operation could be expensive for a
+   *  %BoundedVector and if it is frequently used the user should
+   *  consider using std::list.
+   */
+  template<
+    typename _InputIterator
+  >
+  typename _Base::iterator
+  insert(
+    typename _Base::const_iterator __position,
+    _InputIterator __first,
+    _InputIterator __last)
+  {
+    if (size() + std::distance(__first, __last) > _UpperBound) {
+      throw std::length_error("Exceeded upper bound");
+    }
+    return _Base::insert(__position, __first, __last);
+  }
+
+  using _Base::erase;
+  using _Base::pop_back;
+  using _Base::clear;
+};
+
+/**
+ *  @brief  Vector equality comparison.
+ *  @param  __x  A %BoundedVector.
+ *  @param  __y  A %BoundedVector of the same type as @a __x.
+ *  @return  True if the size and elements of the vectors are equal.
+ *
+ *  This is an equivalence relation.  It is linear in the size of the
+ *  vectors.  Vectors are considered equivalent if their sizes are equal,
+ *  and if corresponding elements compare equal.
+*/
+template<typename _Tp, std::size_t _UpperBound, typename _Alloc>
```

Comment:
Should it be possible to compare `BoundedVector`'s which have different `_UpperBound`'s? Currently I think this code would not compile:

```c++
BoundedVector<int, 32> bv1;
BoundedVector<int, 64> bv2;
bool is_equal = bv1 == bv2;
```

Maybe I'm wrong about that (I haven't tried it). I suppose the same is true if the allocators are different? Maybe that's fine.